src/defrag.c

```diff
@@ -935,6 +935,61 @@ void* defragStreamConsumerGroup(raxIterator *ri, void *privdata) {
     return cg;
 }
 
+/* Defrag a single idmpProducer's dict and linked list entries. */
+static void defragIdmpProducer(idmpProducer *producer) {
+    if (producer->idmp_dict == NULL) return;
+
+    dict *newdict = dictDefragTables(producer->idmp_dict);
+    if (newdict)
+        producer->idmp_dict = newdict;
+
+    idmpEntry **prevnext = &producer->idmp_head;
+    idmpEntry *entry = producer->idmp_head;
+
+    while (entry != NULL) {
+        idmpEntry *next = entry->next;
+        idmpEntry *newentry = activeDefragAllocWithoutFree(entry);
+        if (newentry) {
+            uint64_t hash = dictGetHash(producer->idmp_dict, entry);
+            dictEntry *de = dictFindByHashAndPtr(producer->idmp_dict, entry, hash);
+            serverAssert(de);
+            dictSetKey(producer->idmp_dict, de, newentry);
```

Comment:
dictFindByHashAndPtr(key was freed after defrag) was used before activeDefragAllocWithoutFree was added, now keys was kept via activeDefragAllocWithoutFree, so it's safe to use dictFind now.
```suggestion
            dictEntry *de = dictFindByHashAndPtr(producer->idmp_dict, entry);
            serverAssert(de);
            dictSetKey(producer->idmp_dict, de, newentry);
```